### ISSUE TITLE
Grease is sometimes repeating values

### ISSUE THREAD

**Issue by blakebyrnes** [refraction-networking/utls#52]
GREASE values are sometimes repeated in the current setup. There are some CDNs and sites that are running something on the backend that is rejecting this, but it happens very infrequently, so it's hard to pickup on. I do not know how to write a unit test for this, so I've been putting off submitting a pr for you. You can see the basic code changes here:
https://github.com/ulixee/utls/commit/3b2896145f6b68ca1cd842e603b6043a2b799761

Basically, the current version doesn't actually check if the resulting value is changed.

**Comment by sergeyfrolov** [comment]
I don't know of a reason why would a remote backend reject a repeated connection with the same GREASE value(?). Are you sure GREASE is the culprit?

**Comment by blakebyrnes** [comment]
The GREASE spec says you're not allowed to repeat values. I know some places like CloudFlare are actually running BoringSSL as their TLS handler, so I'm guessing it's coming from something like this.

https://tools.ietf.org/html/rfc8701
```  Implementations advertising GREASE values SHOULD select them at
   random.  This is intended to encourage implementations to ignore all
   unknown values rather than any individual value.  Implementations
   MUST honor protocol specifications when sending GREASE values.  For
   instance, Section 4.2 of [RFC8446] forbids duplicate extension types
   within a single extension block.  Implementations sending multiple
   GREASE extensions in a single block must therefore ensure the same
   value is not selected twice.```

I know the problem went away when I changed it to ensure the GREASE truly never repeats itself, I stopped getting the random errors.

**Comment by blakebyrnes** [comment]
This latter part was also a theory - that servers are simply seeing duplicated extensions, and rejecting because of that.

**Comment by blakebyrnes** [comment]
I just re-read your question. This is not in a repeated connection. This is repeated GREASE values within the same ClientHello request. The code is attempting to ensure they're different, but there's a subtle bug where they can occasionally repeat.

**Comment by sergeyfrolov** [comment]
I see.
Commented in your fork on what I think could be the right fix.
